### DISCUSSION TITLE
Request.body in POST request is empty

### DISCUSSION THREAD

**Discussion by edoardohorse** [vercel/vercel#6628] [Help] [answered]
Hi, I'm having an issue with POST request.
I implemented this [example](https://vercel.com/guides/handling-node-request-body?query=post).

1) If the request start from inside of the app itself, it doesn't receive `req.body` at all! 
![image](https://user-images.githubusercontent.com/17524390/130573626-8d3dceed-9c49-4ef4-85c3-2972a5f41776.png)
2) But if the same request come from my local machine using POSTMAN/cURL, it works! It can read the JSON that I'm sending to the Serveless function. 
![image](https://user-images.githubusercontent.com/17524390/130574095-68592e77-f7e8-46c3-bc40-cd38468432bf.png)

## Code
![image](https://user-images.githubusercontent.com/17524390/130574158-cd5a8327-a470-4556-b63b-0abead131998.png)

## Guessing problem
I think that Vercel can't get request from the same-origin, but can accept the ones from different origin. But it doesn't make sense. I didn't setted no CORS (vercel.json/ next.config.json) file. I tried to set them, with [next.config.json](https://vercel.com/support/articles/how-to-enable-cors?query=cors#enabling-cors-in-a-next.js-app) or [vercel.json](https://vercel.com/support/articles/how-to-enable-cors?query=cors#enabling-cors-using-vercel.json) without success.

Can I get some help please 🙏

**Comment by leerob** [comment, marked as answer]
I've just confirmed this is working. I created a function like shown:

```js
// api/test.js

export default async function handler(req, res) {
  const { body } = req;
  return res.send(`Hello ${body.name}, you just parsed the request body!`);
}
```
Then ran `vercel dev`, which works locally:
<img width="623" alt="CleanShot 2021-08-24 at 14 15 52@2x" src="https://user-images.githubusercontent.com/9113740/130676953-ad5a2bb6-1db9-43fb-ad37-f93800a1b789.png">
As well as when deployed after running `vercel`:
<img width="806" alt="CleanShot 2021-08-24 at 14 20 14@2x" src="https://user-images.githubusercontent.com/9113740/130676935-d2824c1b-2963-49d8-a6cb-0ccfa14ce9db.png">

**Reply by edoardohorse** [reply]
Yes in this way, works for me too. I can send request from the outside (different origin), but that problem is that I can't send a request from the same origin. Imagine of using https://node-headers.vercel.app/ and I send a request to api/test with fetch API, the request is accepted by the server but the body of req is empty.

**Comment by aecorredor** [comment]
I can confirm this is happening to me too. My use case:

1. I'm using Moralis Webhooks (https://docs.moralis.io/moralis-server/cloud-code/webhooks) which basically just call my Vercel function.
2. When I'm running the server locally (using `next dev` and ngrok) I get the request body and everything works.
3. When deployed to Vercel, the body is empty. I noticed that Moralis is not sending `content-length` on the request headers when it calls my Vercel endpoint. That may be affecting the body not getting parsed correctly by `body-parser`. I reproduced the behavior in postman by taking out the `content-length` header (which postman generates automatically).

I'm guessing Vercel expects the `content-length` to be present in order for `body-parser` to work? 

Is there a way to hook into the request lifecycle and force it to parse the body even if the `content-length` header is not present?

**Comment by congtrung6391** [comment]
I also have faced similar when using `fetch` in next-auth. Replacing `fetch` by `axios` help me resolve it.
